### PR TITLE
Forbid `PretrainedConfig` from saving `generate` parameters; Update deprecations in `generate`-related code 🧹 

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

See title :) I did a quick scan and update over things that fulfilled the following criteria:
- had deprecation messages
- were related to what I've been working on

______________________________________________

Key changes: 
1. Saving a model config with parameters meant to be in the generation config is no longer allowed
2. However, at model saving time (which saves related configs), move those parameters to the generation config before potentially complaining about it. This is to prevent fine-tuning jobs from failing after significant compute $ is spent.

### CODE REVIEW

File: src/transformers/commands/pt_to_tf.py
Comment (gante) [inline src/transformers/commands/pt_to_tf.py:15]: TL;DR disabled TF conversion (the warning said it would be removed in v4.43), but kept the error message which redirects to the safetensor conversion guide :)

File: src/transformers/modeling_utils.py
Comment (gante) [inline src/transformers/modeling_utils.py:2554]: NOTE: the check removed here is redundant with the check that gets upgraded from warning to exception in this PR. We call `config.save_pretrained()` above in L2552, which runs the upgraded check 👼

File: src/transformers/models/bloom/modeling_bloom.py
Comment (gante) [inline src/transformers/models/bloom/modeling_bloom.py:696]: found in my quick scan for deprecations (CMD+F on "v4.41") -- let's use our latest docs to guide our users

**Review by amyeroberts — APPROVED** [review]
Killing old code 🤩

File: src/transformers/configuration_utils.py
Comment (gante) [inline src/transformers/configuration_utils.py:1090]: This situation was not encoded before this PR (and caused tests to fail, god bless tests 🙏 )

File: src/transformers/modeling_utils.py
Comment (gante) [inline src/transformers/modeling_utils.py:2552]: Instead of warning (before this PR) / raising an exception (this PR if L2552-2563 were not added) when `model.config` holds things that should be in `model.generation_config` AND were explicitly set by the user (e.g. `model.config.top_k = 5`), now we automagically move misplaced items ✨ 

Automagically moving things is not usually recommended, but this is a special case: `model.save_pretrained` is often run at the end of training jobs, we really want to avoid expensive crashes

File: src/transformers/models/llama/modeling_llama.py
Comment (gante) [reply src/transformers/models/llama/modeling_llama.py:961]: We may want to use `DynamicCache.from_legacy_cache` in non-deprecated ways :p for instance, some optimum integrations use it

**Comment by gante** [comment]
(@amyeroberts don't review yet, I'm going through failed tests -- mostly cases where we are explicitly setting `generate` arguments in `model.config`, precisely what we don't want users to do)

**Comment by amyeroberts** [comment]
@gante I'm going to unsubscribe to avoid notifications for each push. Just ping / @ me to let me know when it's ready for review again!

File: .circleci/create_circleci_config.py
Comment (gante) [inline .circleci/create_circleci_config.py:331]: (the default is 12 workers)
Without this, we're having a concurrency error downloading NLTK data ([e.g.](https://app.circleci.com/pipelines/github/huggingface/transformers/100691/workflows/56c6d291-9d7a-481d-8e29-dcbf96e2bc5d/jobs/1341285)). If we track down that exception, we end up in the NLTK repo, where a maintainer suggests not having concurrency as a fix

File: src/transformers/configuration_utils.py
Comment (gante) [inline src/transformers/configuration_utils.py:91]: Instead of documenting `generate` parameters in `PretrainedConfig`, which is deprecated, show a warning in the docs pointing towards `GenerationConfig`

File: src/transformers/generation/tf_logits_process.py
Comment (gante) [inline src/transformers/generation/tf_logits_process.py:526]: We disallowed storing custom generate flags in the model config -> we can no longer set custom generate parameterization in the model tester [which is not necessarily bad, all models should be able to generate with their default parameters!] -> we couldn't hardcode our way out of edge cases :)

This change fixes the edge case where we set a token that's not in the vocab to be suppressed. The alternative would be to reconfigure the tester to have a vocab size larger than the default value for this parameter, >50k tokens. Given that it is pretty harmless, I went with this :)

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:1371]: this was actually incorrect: we want to detect changes against the default model config, not against the default generation parameters

(To recap: if a user updates the model config and the generation config was automatically generated from it, throw a warning)

File: src/transformers/models/encoder_decoder/configuration_encoder_decoder.py
Comment (gante) [inline src/transformers/models/encoder_decoder/configuration_encoder_decoder.py:78]: I added a try/except for composite models, catching `ValueError`, which led to to this discovery :) You'll see this pattern fixed in other composite models.

File: tests/models/bart/test_modeling_bart.py
Comment (gante) [inline tests/models/bart/test_modeling_bart.py:130]: If we go check the PR that added these, we can read that this change was added to maybe fix a bug that happens sporadically 👀 

I don't agree with the description of the problem given our current code base, I believe the root cause has been addressed. Furthermore, it's good to remove fudge factors 😉

File: tests/models/blenderbot/test_modeling_blenderbot.py
Comment (gante) [inline tests/models/blenderbot/test_modeling_blenderbot.py:371]: no clue why this would be here 🤷

File: tests/models/decision_transformer/test_modeling_decision_transformer.py
Comment (gante) [inline tests/models/decision_transformer/test_modeling_decision_transformer.py:44]: No idea why this was here :)

File: tests/models/mamba2/test_modeling_mamba2.py
Comment (gante) [inline tests/models/mamba2/test_modeling_mamba2.py:232]: this test was failing on `main` -- perhaps because of concurrent merges?

File: tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py
Comment (gante) [inline tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py:418]: Failing on `main` too.

I suspect this didn't get detected in the (recent) PR that added it because of our test fetched, which only checks the main models on these wide changes (?)

**Comment by gante** [comment]
@amyeroberts Ready 🙌 

Notes:
1. saving a config with custom generate parameters now creates an exception -> all testers that relied on setting custom generation parameters to work around edge cases failed [because we pass the tester args to the model config] -> this PR fixes edge cases/updates tests 💪 
2. touching many files triggers a lot of test runs. A few unrelated broken tests were found (and fixed) in the process
3. I've added GH comments on details that are both not useful for the future codebase AND not obvious why the change was made
4. an unrelated pipeline test is failing due to timeout, triggering a new run is not making CI green. I'll need you to merge when you're happy with the PR :D

File: src/transformers/models/audio_spectrogram_transformer/configuration_audio_spectrogram_transformer.py
Comment (amyeroberts) [inline src/transformers/models/audio_spectrogram_transformer/configuration_audio_spectrogram_transformer.py:128]: I'm not sure I understand this - why would there be checks run on a config -> generation_config if the model isn't compatible with generate? 

It seems like a bad pattern to need a generate specific method added to config if the model can't be used for generation

**Review by amyeroberts — COMMENTED** [review]
Thanks for handling all of these updates! 

A few questions / comments but overall looks good 🤗

File: tests/utils/test_modeling_utils.py
Comment (gante) [reply tests/utils/test_modeling_utils.py]: @amyeroberts that is a great question -- what gets written in `model.config` in this case. I agree with your comment (`If it's its old value, I'm worried this could cause confusion`), which was the actual behavior in the commit you reviewed, so I've made the following adjustments:
1. Moving a parameter from `model.config` to `model.generation_config` sets that parameter to `None` in `model.config`.
2. Storing a generation parameter as `None` in `model.config` is now accepted (even if it is not the default value);
3. Updated the warning message, explaining to the user why they are seeing it
4. The test now also checks that the warning is thrown. It is also commented, so it should be simple to understand what should be happening.
🤗

File: tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py
Comment (gante) [reply tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py]: https://github.com/huggingface/transformers/issues/32913

(The issue catches a wider net than skipping `test_inputs_embeds_matches_input_ids_with_generate`, as this skip is merely a symptom of the root problem :) )

File: .circleci/create_circleci_config.py
Comment (gante) [reply .circleci/create_circleci_config.py:331]: NTLK issue: https://github.com/nltk/nltk/issues/3248 

Workarounds would all consist of running this part of the CI job without parallelism or adding locks -- IMO non-trivial stuff. Given that it blocks any PR that touches code that might be used in TF, such as `configuration_utils.py`, it's pretty disruptive 😢 I think adding this band-aid and chasing a better solution in parallel would be the best 🤗

File: src/transformers/generation/tf_logits_process.py
Comment (gante) [reply src/transformers/generation/tf_logits_process.py:526]: @amyeroberts We can increase the vocabulary size of the tester to the original vocabulary size of 50+k tokens (instead of the dummy vocab size of 100).

Happy to do that instead :)

[Note: given TF's indexing, where GPU+no XLA gladly accepts out-of-bounds indexes, this logic actually streamlines the behavior across devices and/or XLA.]

File: src/transformers/models/audio_spectrogram_transformer/configuration_audio_spectrogram_transformer.py
Comment (gante) [reply src/transformers/models/audio_spectrogram_transformer/configuration_audio_spectrogram_transformer.py:128]: I understand and share your pain 😬 Allow me to elaborate.

The root issue is in the model config. A model config is agnostic to whether the model can or cannot generate, unless we hardcode it in the config class or access external objects. 

When we call `config.save_pretrained`, we don't know whether we should or should not check generation parameters. Since we want to move all generation-related logic to `GenerationConfig`, whose one of the goals is to avoid clashes like this and allow proper validation, we have to assume all configs may be holding generate parameters. The current logic enforces no custom generate parameters are held in the model config, so no new models parameterize `generate` through the model config.

AST is the exception. It doesn't support `generate`, but it has a model parameter whose name is also used in `generate` (`max_length`). If a user creates a model with a custom `max_length`, which should be allowed, the current `save_pretrained` would crash because it thinks it has a custom `generate` parameter. This workaround effectively skips the `generate` checks, since we know it can't generate.

Between all three alternatives I could see, this seemed the best solution -- especially since it is a temporary fix, to be removed after we untangle the two configs (v5.0?)

Considered fixes:
- skip generate-related checks on this specific class (implemented)
- add a parameter to ALL model config classes to store whether the model can generate
- add an argument to `config.save_pretrained` to check whether to check `generate` parameters, receive `model.can_generate()` when called from `model.save_pretrained`

**Comment by gante** [comment]
(CI was failing, rebased with main to check if that was the issue)

File: .circleci/create_circleci_config.py
Comment (ydshieh) [reply .circleci/create_circleci_config.py:331]: I don't know what happen recently, but we don't have this error before.

If we decide to have this quick (temp) fix, it might be a good idea to have a small PR for only this part.

Comment (gante) [reply .circleci/create_circleci_config.py:331]: Opening a separate PR :)

File: tests/utils/test_modeling_utils.py
Comment (amyeroberts) [inline tests/utils/test_modeling_utils.py:1603]: Nice :)

File: src/transformers/models/audio_spectrogram_transformer/configuration_audio_spectrogram_transformer.py
Comment (amyeroberts) [reply src/transformers/models/audio_spectrogram_transformer/configuration_audio_spectrogram_transformer.py:128]: Make sense - thanks for taking the time to explain!

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating and taking the time to answer all my qs. 

As this touches so many models and core logic, we should ideally do a `[run_all]` run of slow tests to make sure everything is tip-top before merge

**Pull request merged**
